server/src/main/java/com/linecorp/centraldogma/server/internal/mirror/MirrorRunner.java

```diff
@@ -0,0 +1,154 @@
+
+public final class MirrorRunner implements SafeCloseable {
+
+    private final ProjectApiManager projectApiManager;
+    private final CommandExecutor commandExecutor;
+    private final File workDir;
+    private final MirroringServicePluginConfig mirrorConfig;
+    private final ExecutorService worker;
+
+    private final Map<MirrorKey, CompletableFuture<MirrorResult>> inflightRequests = new ConcurrentHashMap<>();
+
+    public MirrorRunner(ProjectApiManager projectApiManager, CommandExecutor commandExecutor,
+                        CentralDogmaConfig cfg, MeterRegistry meterRegistry) {
+        this.projectApiManager = projectApiManager;
+        this.commandExecutor = commandExecutor;
+        // TODO(ikhoon): Periodically clean up stale repositories.
+        workDir = new File(cfg.dataDir(), "_mirrors_manual");
+        MirroringServicePluginConfig mirrorConfig =
+                (MirroringServicePluginConfig) cfg.pluginConfigMap()
+                                                  .get(MirroringServicePluginConfig.class);
+        if (mirrorConfig == null) {
+            mirrorConfig = MirroringServicePluginConfig.INSTANCE;
+        }
+        this.mirrorConfig = mirrorConfig;
+
+        final ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(
+                0, mirrorConfig.numMirroringThreads(),
+                // TODO(minwoox): Use LinkedTransferQueue when we upgrade to JDK 21.
+                60, TimeUnit.SECONDS, new LinkedBlockingQueue<>(),
+                new DefaultThreadFactory("mirror-api-worker", true));
+        threadPoolExecutor.allowCoreThreadTimeOut(true);
+        worker = ExecutorServiceMetrics.monitor(meterRegistry, threadPoolExecutor,
+                                                "mirrorApiWorker");
+    }
+
+    public CompletableFuture<MirrorResult> run(String projectName, String mirrorId) {
+        // If there is an inflight request, return it to avoid running the same mirror task multiple times.
+        return inflightRequests.computeIfAbsent(new MirrorKey(projectName, mirrorId), this::run);
+    }
+
+    private CompletableFuture<MirrorResult> run(MirrorKey mirrorKey) {
+        try {
+            final CompletableFuture<MirrorResult> future =
+                    metaRepo(mirrorKey.projectName).mirror(mirrorKey.mirrorId).thenApplyAsync(mirror -> {
+                        if (!mirror.enabled()) {
+                            throw new MirrorException("The mirror is disabled: " + mirrorKey);
+                        }
+
+                        return mirror.mirror(workDir, commandExecutor,
+                                             mirrorConfig.maxNumFilesPerMirror(),
+                                             mirrorConfig.maxNumBytesPerMirror());
+                    }, worker);
+            // Remove the inflight request when the mirror task is done.
+            future.handleAsync((unused0, unused1) -> inflightRequests.remove(mirrorKey));
```

Comment:
I didn't use `handle` intentionally because `inflightRequests.remove` could be called recursively inside `inflightRequests.computeIfAbsent()`. I wasn't confident that all supported JDK versions were safe for the operation.


Let me specify `worker` for the job.